File: sdk/runanywhere-swift/Sources/RunAnywhere/Public/Models/GenerationOptions.swift
import Foundation

/// Options for text generation
public struct RunAnywhereGenerationOptions {
    /// Maximum number of tokens to generate
    public let maxTokens: Int

    /// Temperature for sampling (0.0 - 1.0)
    public let temperature: Float

    /// Top-p sampling parameter
    public let topP: Float

/// Enable real-time tracking for cost dashboard
    public let enableRealTimeTracking: Bool

    /// Stop sequences
    public let stopSequences: [String]

    /// Seed for reproducible generation
    public let seed: Int?

    /// Enable streaming mode
    public let streamingEnabled: Bool

    /// Token budget constraint (for cost control)
    public let tokenBudget: TokenBudget?

    /// Framework-specific options
    public let frameworkOptions: FrameworkOptions?

    /// Preferred execution target
    public let preferredExecutionTarget: ExecutionTarget?

    /// Structured output configuration (optional)
    public let structuredOutput: StructuredOutputConfig?

    /// System prompt to define AI behavior and formatting rules
    public let systemPrompt: String?

    /// Initialize generation options
    /// - Parameters:
    ///   - maxTokens: Maximum tokens to generate (default: 100)
    ///   - temperature: Sampling temperature (default: 0.7)
    ///   - topP: Top-p sampling (default: 1.0)
///   - enableRealTimeTracking: Enable real-time cost tracking (default: true)
    ///   - stopSequences: Stop generation at these sequences (default: empty)
    ///   - seed: Optional seed for reproducibility
    ///   - streamingEnabled: Enable streaming mode (default: false)
    ///   - tokenBudget: Token budget constraints
    ///   - frameworkOptions: Framework-specific options
    ///   - preferredExecutionTarget: Preferred execution target
    ///   - structuredOutput: Structured output configuration (optional)
    ///   - systemPrompt: System prompt for AI behavior (optional)
    public init(
        maxTokens: Int = 100,
        temperature: Float = 0.7,
        topP: Float = 1.0,
enableRealTimeTracking: Bool = true,
        stopSequences: [String] = [],
        seed: Int? = nil,
        streamingEnabled: Bool = false,
        tokenBudget: TokenBudget? = nil,
        frameworkOptions: FrameworkOptions? = nil,
        preferredExecutionTarget: ExecutionTarget? = nil,
        structuredOutput: StructuredOutputConfig? = nil,
        systemPrompt: String? = nil
    ) {
        self.maxTokens = maxTokens
        self.temperature = temperature
        self.topP = topP
self.enableRealTimeTracking = enableRealTimeTracking
        self.stopSequences = stopSequences
        self.seed = seed
        self.streamingEnabled = streamingEnabled
        self.tokenBudget = tokenBudget
        self.frameworkOptions = frameworkOptions
        self.preferredExecutionTarget = preferredExecutionTarget
        self.structuredOutput = structuredOutput
        self.systemPrompt = systemPrompt
    }
}

File: sdk/runanywhere-android/src/main/java/com/runanywhere/sdk/models/GenerationOptions.kt
package com.runanywhere.sdk.models

/**
 * Options for text generation
 */
data class GenerationOptions(
    /**
     * Maximum number of tokens to generate
     */
    val maxTokens: Int = 100,
    
    /**
     * Temperature for sampling (0.0 - 1.0)
     */
    val temperature: Float = 0.7f,
    
    /**
     * Top-p sampling parameter
     */
    val topP: Float = 1.0f,
    
    /**
     * Context for the generation
     */
    val context: Context? = null,
    
    /**
     * Enable real-time tracking for cost dashboard
     */
    val enableRealTimeTracking: Boolean = true,
    
    /**
     * Stop sequences
     */
    val stopSequences: List<String> = emptyList(),
    
    /**
     * Seed for reproducible generation
     */
    val seed: Int? = null,
    
    /**
     * Enable streaming mode
     */
    val streamingEnabled: Boolean = false,
    
    /**
     * Token budget constraint (for cost control)
     */
    val tokenBudget: TokenBudget? = null,
    
    /**
     * Framework-specific options
     */
    val frameworkOptions: FrameworkOptions? = null,
    
    /**
     * Preferred execution target
     */
    val preferredExecutionTarget: ExecutionTarget? = null
)

/**
 * Context for maintaining conversation state
 */
data class Context(
    /**
     * Previous messages in the conversation
     */
    val messages: List<Message> = emptyList(),
    
    /**
     * System prompt override
     */
    val systemPrompt: String? = null,
    
    /**
     * Maximum context window size
     */
    val maxTokens: Int = 2048
)

/**
 * Message in a conversation
 */
data class Message(
    /**
     * Role of the message sender
     */
    val role: Role,
    
    /**
     * Content of the message
     */
    val content: String,
    
    /**
     * Timestamp
     */
    val timestamp: Long = System.currentTimeMillis()
) {
    enum class Role {
        USER,
        ASSISTANT,
        SYSTEM
    }
}

/**
 * Token budget for cost control
 */
data class TokenBudget(
    val maxTokens: Int,
    val costLimit: Double? = null
)

/**
 * Framework-specific options
 */
data class FrameworkOptions(
    val tensorFlowLiteOptions: TensorFlowLiteOptions? = null,
    val onnxOptions: OnnxOptions? = null,
    val llamaCppOptions: LlamaCppOptions? = null
)

/**
 * TensorFlow Lite specific options
 */
data class TensorFlowLiteOptions(
    val useNNAPI: Boolean = false,
    val useGPU: Boolean = false,
    val numThreads: Int = 4
)

/**
 * ONNX specific options
 */
data class OnnxOptions(
    val executionProvider: String = "CPUExecutionProvider",
    val graphOptimizationLevel: Int = 99
)

/**
 * Llama.cpp specific options
 */
data class LlamaCppOptions(
    val nCtx: Int = 2048,
    val nThreads: Int = 4,
    val nGpuLayers: Int = 0
) 